File: workstations.java
import java.util.Scanner;
import java.util.Arrays;
import java.util.ArrayList;

public class workstations {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int r_len = sc.nextInt();
        int activity = sc.nextInt();
        
        int[] arrives = new int[r_len];
        int[] leaves = new int[r_len];
        
        for(int r = 0; r < r_len; ++r) {
            int arrive = sc.nextInt();
            arrives[r] = arrive;
            
            int stays = sc.nextInt();
            leaves[r] = arrive + stays;
        }
        
        Arrays.sort(arrives);
        Arrays.sort(leaves);
        int result = 0;
        
        for(int r = 0; r < r_len; ++r) {
            int ar = arrives[r];
            
            for(int i = 0; i < r_len; ++i) {
                if(leaves[i] == -1) {
                    continue;
                }
                
                if(leaves[i] <= ar && leaves[i] + activity >= ar) {
                    ++result;
                    leaves[i] = -1;
                    break;
                }
            }
        }
        System.out.println(result);
        sc.close();
    }
}

File: workstations.cpp
#include <cstdio>
#include <algorithm>

int arrives[300000];
int leaves[300000];

int main() {
    int r_len, activity;
    scanf("%d %d", &r_len, &activity);
    
    for(int r = 0; r < r_len; ++r) {
        int arrive;
        scanf("%d", &arrive);
        
        arrives[r] = arrive;
        
        int stays;
        scanf("%d", &stays);
        leaves[r] = arrive + stays + activity;
    }
    
    std::sort(std::begin(arrives), std::begin(arrives) + r_len);
    std::sort(std::begin(leaves), std::begin(leaves) + r_len);
    
    int result = 0;
    int l_len = r_len;
    for(int r = 0; r < r_len; ++r) {
        int ar = arrives[r];
        
        auto it = std::lower_bound(std::begin(leaves), std::begin(leaves) + l_len, ar);
        
        if(*it >= ar && *it - activity <= ar) {
            ++result;
            
            int index = std::distance(std::begin(leaves), it);
            
            for(int i = index; i < l_len; ++i) { 
                std::swap(leaves[i], leaves[i + 1]);
            }
            --l_len;
        }
    }
    printf("%d\n", result);
}